### DISCUSSION TITLE
x-for dynamic dropdown + wire:model

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#3562] [1. Help]
I have a blade component that gets data from an API using javascript to create options in a dropdown menu.  I use `x-for` to loop over all entries and create them here.

```blade
<x-input.select x-show="!error" {{ $attributes->whereStartsWith('wire:model') }}>
    <option value="null" disabled>Please select a printer..</option>
    <template x-for="(printer, index) in printers" :key="index">
        <option x-bind:value="printer.deviceId" x-text="printer.name"></option>
    </template>
</x-input.select>
```

I've left out the javascript that populates `printers`, but this works, brilliantly. What might be important though is that it's an async function that loads the printers. 

We use the blade component in our Livewire components like:

```blade
<x-select-printer wire:model="user.default_printer" id="default_printer" />
```

The above doesn't work for me. Here's what I do:

- I load up the page with the blade component: it says `Please select a printer..` (as expected)
- I select a printer, hit save, and the livewire component stores it in the database (as expected)
- When I refresh the Livewire component it shows the first printer in the list, even though this is not the printer I selected and was stored in the database.

I'm guessing that Livewire has done its magic, before `x-for` runs. But it might be something else.

Do I have to do something extra when combining `x-for` with `wire:model`? What am I doing wrong? :)

**Comment by joshhanley** [comment]
@ju5t yeah, Livewire will probably already have been instantiated before the results are available, so I don't think that will work.

Can you do the initial API call from Livewire instead?

If you need it to work that way, I probably wouldn't be using a select element, instead I'd be building a custom dropdown, where can controlled which is selected in Alpine.

Hope this helps!

**Reply by ju5t** [reply]
@joshhanley thought as much. I will try to find a solution with `entangle`.

The API call can't be moved to Livewire. We have a small Electron app that boots up a REST API on the users local machine. It allows us to fetch a list of their printers and print a pdf (e.g. packing lists) without a confirmation popup to a specific printer of them. So I have to use javascript, it's the only way to query localhost :)

Thanks for the help/confirmation of my idea, as usual!

**Reply by joshhanley** [reply]
@ju5t oh interesting! Right, so it needs to make the call locally using JS, makes sense. Yeah I'd be looking at a custom dropdown, or finding a way to "refresh" the select element so it picks up the latest changes.

Basically, if Livewire can't find the option that matches its current value, it will just select the first in the list, which is why, once it's loaded you're seeing the first printer selected. I'm not sure how you could get Livewire to detect that change.

Maybe, actually, now thinking about it, if you `@entangle` the value into Alpine, then, when you are looping and outputting your options, you can manually output `selected` on the one that matches, and maybe that would force the UI to have the correct one selected. Because ultimately, LW isn't wrong in the value it holds, it's just the HTML that is wrong.

I actually think just outputting selected by tracking it with `@entangle` may fix it.

No worries 🙂

**Reply by joshhanley** [reply]
If outputting selected doesn't work, use JS and get the select element and either set the `selectedIndex` on it, or the `value`, to what Livewire has, and that should get the UI to marry up with what Livewire has.

**Reply by ju5t** [reply]
In my data method I added:

`selected: @entangle($attributes->wire('model')),`

And instead of binding the select, I switched it to:

`x-model="selected"`

Then, in the loop, on the option element, I added:

`x-bind:selected="selected === printer.deviceId"`

Initially, I didn't expect I had to add the `x-bind:selected` to the loop as I kind of expected `x-model` to take care of that. But, this is probably a bit more expressive and easier to understand for future me, so pretty happy nonetheless. I know without the javascript it probably doesn't make a lot of sense, but it works :)

**Reply by joshhanley** [reply]
@ju5t nice! Hmm yeah actually you'd think the `x-model` would, but at least it works! 🙂
